### PR TITLE
AYS long running jobs initial implementation

### PR DESCRIPTION
**Pull request by xmonader**
Experimental FR for implementing long running jobs as coroutines.

Fixing https://github.com/Jumpscale/ays9/issues/105

### CODE REVIEW

File: JumpScale9AYS/ays/lib/Recurring.py [hunk 1]
```diff
@@ -54,6 +53,30 @@ def stop(self):
...
+    def __init__(self, service, action, loop=None):
+        super().__init__(service=service, action=action, period=None, loop=loop)
+        self.logger.info("Created long task {} of service {}".format(action, service))
+
+    async def _run(self):
+        try:
+            # create job
+            self._job = self.service.getJob(actionName=self.action)
+            self.logger.info("in run long task")
+            # compute how long we need to sleep before next execution
```
Comment (zaibon) [inline JumpScale9AYS/ays/lib/Recurring.py]: ?? remaining of the copy/paste ?

File: JumpScale9AYS/ays/lib/Recurring.py [hunk 2]
```diff
@@ -54,6 +53,30 @@ def stop(self):
...
+            # wait for right time
+            await asyncio.sleep(1)
```
Comment (zaibon) [inline JumpScale9AYS/ays/lib/Recurring.py]: Why sleep here ?

File: docs/LongRunningJobs.md [hunk 3]
```diff
@@ -0,0 +1,38 @@
+# Long running jobs
+
+AYS is written using asyncio technique to achieve concurrency.
+
+## Introduction
+You can have long running tasks aka `recurring tasks` and it'll serve you well to a limit as it uses `loop.run_in_executor` to not block `ays main thread`, if you are having too much long running jobs maybe you need to use `LongRunningTask` concept.
+
+## How it Works
+You basically define a coroutine to be `injected` in the main loop, which is `DANGEROUS`, because again, you will end in a `blocked AYS server`.
```
Comment (zaibon) [inline docs/LongRunningJobs.md]: This needs to be improved

File: docs/LongRunningJobs.md [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+
+
+## How to define a `Long Running Task`
+
+in `actions.py` of your service.
```
Comment (zaibon) [inline docs/LongRunningJobs.md]: Write a meaning full example.

File: JumpScale9AYS/jobcontroller/Job.py [hunk 5]
```diff
@@ -315,7 +315,7 @@ def executeInProcess(self):
         """
         return self.execute()
 
-    def execute(self):
+    def execute(self, sameloop=False):
```
Comment (zaibon) [inline JumpScale9AYS/jobcontroller/Job.py]: Why do we need this argument ?
You already know if the job is long running or not, so you don't need an extra argument to decide how to start the job.

File: JumpScale9AYS/ays/lib/model.capnp [hunk 6]
```diff
@@ -50,6 +50,8 @@ struct Actor {
     state @4 :ActionState;
     isJob @5 :Bool;
     timeout @6 :UInt32;
+    longjob @7 :Bool = false;
```
Comment (zaibon) [inline JumpScale9AYS/ays/lib/model.capnp]: you need to change the ID of the file when you update the schema

File: JumpScale9AYS/jobcontroller/Job.py [hunk 7]
```diff
@@ -327,8 +327,19 @@ def execute(self):
         if self.model.dbobj.debug is False:
             self.model.dbobj.debug = self.sourceLoader.source.find('ipdb') != -1 or \
                                      self.sourceLoader.source.find('IPython') != -1
+        if self.service.model.actions[self.action.dbobj.name].longjob is False:
+            self._future = self._loop.run_in_executor(None, self.method, self)
+        else:
+            # THIS FEATURE IS VERY DANGEROUS: USE with CARE or you END UP with a blocked AYS.
```
Comment (zaibon) [inline JumpScale9AYS/jobcontroller/Job.py]: Please explain how you could end up with a blocled AYS.
No where you explain that the code needs to be fully async.

File: JumpScale9AYS/ays/lib/Service.py [hunk 8]
```diff
@@ -37,14 +38,16 @@ def init_from_model(cls, aysrepo, model):
...
         return self
 
     @classmethod
     def init_from_fs(cls, aysrepo, path, context=None):
         self = cls(aysrepo=aysrepo)
         self._loadFromFS(path)
         self.aysrepo.db.services.services[self.model.key] = self
-        self._ensure_recurring()
+        self._ensure_recurring
+        self._ensure_longjobs()
```
Comment (ahussein) [inline JumpScale9AYS/ays/lib/Service.py]: what is this ? why did you stop calling the _ensure_recurring...this is now just a reference to the function not a function call! am I missing something ?

**Comment by ahussein** [comment]
We migrated the core tests for ays from 8.2 version and now they are in master branch and run everytime we create a new PR....I think extending the tests for new features is a must. At least we need to add a basic test for the new feature before merging it in.

File: JumpScale9AYS/ays/lib/Service.py (hunk 8 above)
Comment (xmonader) [inline JumpScale9AYS/ays/lib/Service.py]: No, it was ruined with update. sorry about that.

File: JumpScale9AYS/ays/lib/model.capnp (hunk 6 above)
Comment (xmonader) [inline JumpScale9AYS/ays/lib/model.capnp]: Yeah, apparently not to ruin the code for the others.

**Comment by xmonader** [comment]
initial test: https://github.com/Jumpscale/ays9/commit/cd8a1c59dd6ab3e90c0fc8d28a731133fe2cb905

**Pull request merged**